clang/test/CodeGenCXX/GH106182.cpp

```diff
@@ -0,0 +1,12 @@
+// RUN: %clang_cc1 -std=c++20 %s -triple %itanium_abi_triple -emit-llvm -o - | FileCheck %s
+
+template <template <class> class S>
+void create_unique()
+  requires (S{0}, true) {}
+
+template <class Fn> struct A {
+  constexpr A(Fn) {};
+};
+
+template void create_unique<A>();
+// CHECK: @_Z13create_uniqueI1AEvvQcmtlT_Li0EELb1E(
```

Comment:
There is a difference compared to GCC
https://compiler-explorer.com/z/1EvTrecvh

```
_Z13create_uniqueI1AEvvQcmtl1SLi0EELb1E
 => void create_unique<A>() requires T{0}, true

_Z13create_uniqueI1AEvvQcmtlT_Li0EELb1E
 => void create_unique<A>() requires S{0}, true
```
Which i guess is because of the improper canonicalisation

Given that this changes looks correct we should land it, and because main is crashing, we should backport it.
So, we should remove the changelog entry and instead add a changelog entry to 19x explaining the breaking mangling change.

We probably want to make sure this is on GCC's radar.
How hard would it be to get back the older behavior? If at all possible we might want to consider putting the new mangling under an ABI flag.

@AaronBallman @zygoloid @rjmccall

Reply:
> Given that this changes looks correct we should land it, and because main is crashing, we should backport it.
So, we should remove the changelog entry and instead add a changelog entry to 19x explaining the breaking mangling change.

I think we've missed the 19.x cycle for this because it involves an ABI break (and final goes out in less than a week, which is not enough bake time for us to land it before 19.0.1): https://llvm.org/docs/HowToReleaseLLVM.html#release-patch-rules

> If at all possible we might want to consider putting the new mangling under an ABI flag.

+1

Reply:
@AaronBallman to clarify the status quo is that this crashes in 19. (and did not crash in 18)

Reply:
> @AaronBallman to clarify the status quo is that this crashes in 19. (and did not crash in 18)

Yup, I knew it was a regression (boo!); if we fixed it a few weeks ago, I think it'd be an easier sell for backporting. My concern is we don't have enough bake time for 19.0.1, and once that goes out we no longer take ABI breaking changes for subsequent releases of 19.x.

> Lastly, keeping the previous behavior under a different ABI flag seems problematic, as we would be also tying these bugs to the flag.

That's the usual point to the flag; we start doing the right thing by default, but we leave an escape hatch for people relying on the previous ABI (despite the bugs) because we don't know who is relying on the old behavior.

Reply:
So, here is an idea.
Should we
 - Land this patch in 20
 - Revert #95202 in 19 - Yes clang will be be broken but hopefully not more than it was in 18? At least it would not crash.

That seems like the cautious approach that avoid regressions/ICE in 19

As for whether we need to preserve the old behavior of mangling in an ABI flag, I'd like the opinion of more people and of what GCC's plan are. I also think we need to wait for feedback on whether it's impacting folks on the 20 cycle. And if people scream we can try to restore the old behavior. No need to carry bugs if no one rely on them

Reply:
I think that's a reasonable path forward (presuming that we only need to revert one commit); that revert has to happen Real Soon™ though so we can get it backported.

Reply:
> @AaronBallman: I think we've missed the 19.x cycle for this because it involves an ABI break (and final goes out in less than a week [...]

The process does have some leeway for further RCs though if there are many (or critical) regressions, c.f. statement from @tru [here](https://discourse.llvm.org/t/llvm-19-release-schedule-and-planning/79828):
> We have previously inserted a 4th RC when there were a lot of outstanding bugs.

Reply:
https://github.com/llvm/llvm-project/pull/106516